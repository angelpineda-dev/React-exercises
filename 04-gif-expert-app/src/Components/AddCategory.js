import React, { useState } from "react";
import PropTypes from "prop-types";

export const AddCategory = ({ setCategories }) => {
  const [inputValue, setInputValue] = useState("");

  // Function to Input
  const handleInputChange = (e) => {
    setInputValue(e.target.value);
  };

  // Function to Form
  const handleSubmit = (e) => {
    e.preventDefault();

    if (inputValue.trim().length > 2) {
      setCategories((cats) => [inputValue, ...cats]);
      setInputValue("");
    } else {
      console.warn("El campo no debe estar vacio o ser mayor a una letra");
    }
  };

  // What this object return?
  return (
    <form onSubmit={handleSubmit}>
      <p>{inputValue}</p>
      <input
        type="text"
        value={inputValue}
        onChange={handleInputChange}
        placeholder="Buscar..."
      />
    </form>
  );
};

// This object is required...
AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired,
};
