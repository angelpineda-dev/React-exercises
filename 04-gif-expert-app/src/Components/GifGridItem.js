import React from "react";
import PropTypes from "prop-types";

export const GifGridItem = ({ title, url }) => {
  return (
    <div className="card animate__animated animate__fadeIn">
      <img src={url} alt={title} />
      <p>{title}</p>
    </div>
  );
};

/* ¿Cómo implementar propTypes para requerir valores?
 - Las PropTypes son utiles para indicar que clase de valor debe tener una variable y más propiedades.
*/

GifGridItem.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
};
