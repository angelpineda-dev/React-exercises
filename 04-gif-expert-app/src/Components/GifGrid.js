import React /* { useState, useEffect } */ from "react";
import PropTypes from "prop-types";
import { useFetchGifs } from "../hooks/useFetchGifs";
import { GifGridItem } from "./GifGridItem";

export const GifGrid = ({ category }) => {
  // making a request from our hook.
  const { data: images, loading } = useFetchGifs(category);

  return (
    // We're returning a fragment
    <>
      <h3>{category}</h3>

      {loading && (
        <p className="animate__animated animate__flash">Loading...</p>
      )}

      <div className="card-flex">
        {images.map((img) => (
          <GifGridItem key={img.id} {...img} />
        ))}
      </div>
    </>
  );
};

GifGrid.propTypes = {
  category: PropTypes.string.isRequired,
};
