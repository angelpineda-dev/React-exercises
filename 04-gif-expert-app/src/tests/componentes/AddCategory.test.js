import React from "react";
import { AddCategory } from "../../Components/AddCategory";
import "@testing-library/jest-dom";
import { shallow } from "enzyme";

describe("Pruebas en <AddCategory />", () => {
  const setCategories = jest.fn();
  let wrapper = shallow(<AddCategory setCategories={setCategories} />);

  beforeEach(() => {
    jest.resetAllMocks();
    wrapper = shallow(<AddCategory setCategories={setCategories} />);
  });

  test("Debe de mostrarse correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("Debe de cambiar la caja de texto", () => {
    let input = wrapper.find("input");
    let value = "Hola Mundo";

    input.simulate("change", { target: { value } });

    expect(wrapper.find("p").text().trim()).toBe(value);
  });

  test("NO debe postear la informaciÃ³n con el submit", () => {
    wrapper.find("form").simulate("submit", { preventDefault: () => {} });

    expect(setCategories).not.toHaveBeenCalled();
  });

  test("Debe de llamar el setCategories y limpiar la caja de texto", () => {
    /* 
      1. simular el inputChange 
      2. simular el submit
      3. setCategories debe de haberse llamado
      4. el valor del input debe de estar ''
    */

    let input = wrapper.find("input");
    let value = "Goku";

    input.simulate("change", { target: { value } });

    // Probando que contenga el valor
    // expect(wrapper.find("p").text().trim()).toBe(value);

    wrapper.find("form").simulate("submit", { preventDefault() {} });

    // Comprobando que se haya llamado el evento submit.
    expect(setCategories).toHaveBeenCalled();
    expect(setCategories).toHaveBeenCalledTimes(1);
    expect(setCategories).toHaveBeenCalledWith(expect.any(Function));

    expect(wrapper.find("input").prop("value")).toBe("");
  });
});
