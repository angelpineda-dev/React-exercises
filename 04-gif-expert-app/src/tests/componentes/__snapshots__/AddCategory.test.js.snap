// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <AddCategory /> Debe de mostrarse correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    placeholder="Buscar..."
    type="text"
    value=""
  />
</form>
`;
